### DISCUSSION TITLE
Export with uppercase extensions?

### DISCUSSION THREAD

**Discussion by nullpointerninja** [RhetTbull/osxphotos#713] [Q&A]
Hi, first of all thanks for this extremely useful tool!

I've been testing it with some RAW+JPG photos and found out it always exports the RAW files with lowercase extension (i.e. cr2 and raf instead of CR2 and RAF). Is it possible to force an export to always use uppercase extensions while keeping the rest of the filename intact?

Thanks!

**Comment by RhetTbull** [comment]
There is no option to explicitly do this but you can accomplish the same outcome by using the `--post-command` option which runs a custom command after every export.  Add the following to your export and all extensions will be converted to uppercase:

`--post-command exported "mv {shell_quote,{filepath}} {shell_quote,{filepath.parent}/{filepath.stem}{filepath.suffix|upper}}"`

Here's how this works:

`--post-command exported` instructs osxphotos to run the command that follows for each exported image.

`mv ` is just the standard `move file` command built into the operating system.

`{shell_quote,{filepath}}` is a statement written in the OSXPhotos [template language](https://github.com/RhetTbull/osxphotos#template-system) that inserts the full path (`{filepath}`) of the exported photo and performs a "shell quote" on the path so the `mv` command will correctly handle the path if it contains special characters or spaces.  e.g. `/Users/rhet/Desktop/export/[2020-08-29] AAF035 (2).jpg` becomes `'/Users/rhet/Desktop/export/[2020-08-29] AAF035 (2).jpg'`  (Note added single quotes)

`{shell_quote,{filepath.parent}/{filepath.stem}{filepath.suffix|upper}}` is another template statement that builds a new path using the parent directory + filename + uppercase version of the suffix (extension) and shell quotes it. So you get something like this as the command that gets executed:

`mv '/Users/rhet/Desktop/export/[2020-08-29] AAF035 (2).jpg' '/Users/rhet/Desktop/export/[2020-08-29] AAF035 (2).JPG'`

The effectively renames each file after export so the extension is uppercase.  

One note: osxphotos uses a database that tracks each exported filename so if you run export again with `--update`, only new or changed files get exported.  To do this, osxphotos needs to track the exported name of each file and the above command will rename the file. However, in this case, `--update` will still work fine.  This is because osxphotos assumes that the destination file system is case-insensitive (the default for macOS file systems) and thus it performs a case-insensitive compare when comparing new files to exported files.  Thus, you can use this technique and still use `--update`.  

This work-around should suffice for now but I'll look at adding `--ext-upper` and `--ext-lower` options to force all extensions to be uppercase or lowercase. 

Note: I recommend you try this first with `--dry-run --verbose` which will print out your command but not run it just to be sure you've got it right.

**Reply by RhetTbull** [reply]
For additional reference, here's the help on `--post-command`:

```
--post-command CATEGORY COMMAND
                                Run COMMAND on exported files of category CATEGORY.  CATEGORY can be one of:
                                exported, new, updated, skipped, missing, exif_updated, touched,
                                converted_to_jpeg, sidecar_json_written, sidecar_json_skipped,
                                sidecar_exiftool_written, sidecar_exiftool_skipped, sidecar_xmp_written,
                                sidecar_xmp_skipped, error. COMMAND is an osxphotos template string, for
                                example: '--post-command exported "echo
                                {filepath|shell_quote} >> {export_dir}/exported.txt"', which appends the
                                full path of all exported files to the file 'exported.txt'. You can run more
                                than one command by repeating the '--post-command'
                                option with different arguments. See Post Command below.

Post Command

You can run commands on the exported photos for post-processing using the '--post-command' option. '--post-
command' is passed a CATEGORY and a COMMAND. COMMAND is an osxphotos template string which will be rendered
and passed to the shell for execution. CATEGORY is the category of file to pass to COMMAND. The following
categories are available:

Category                  Description
exported                  All exported files
new                       When used with '--update', all newly exported files
updated                   When used with '--update', all files which were previously exported but updated
                          this time
skipped                   When used with '--update', all files which were skipped (because they were
                          previously exported and didn't change)
missing                   All files which were not exported because they were missing from the Photos
                          library
exif_updated              When used with '--exiftool', all files on which exiftool updated the metadata
touched                   When used with '--touch-file', all files where the date was touched
converted_to_jpeg         When used with '--convert-to-jpeg', all files which were converted to jpeg
sidecar_json_written      When used with '--sidecar json', all JSON sidecar files which were written
sidecar_json_skipped      When used with '--sidecar json' and '--update', all JSON sidecar files which were
                          skipped
sidecar_exiftool_written  When used with '--sidecar exiftool', all exiftool sidecar files which were written
sidecar_exiftool_skipped  When used with '--sidecar exiftool' and '--update, all exiftool sidecar files
                          which were skipped
sidecar_xmp_written       When used with '--sidecar xmp', all XMP sidecar files which were written
sidecar_xmp_skipped       When used with '--sidecar xmp' and '--update', all XMP sidecar files which were
                          skipped
error                     All files which produced an error during export

In addition to all normal template fields, the template fields '{filepath}' and '{export_dir}' will be
available to your command template. Both of these are path-type templates which means their various parts can
be accessed using the available properties, e.g. '{filepath.name}' provides just the file name without path
and '{filepath.suffix}' is the file extension (suffix) of the file. When using paths in your command template,
it is important to properly quote the paths as they will be passed to the shell and path names may contain
spaces. Both the '{shell_quote}' template and the '|shell_quote' template filter are available for this
purpose.  For example, the following command outputs the full path of newly exported files to file 'new.txt':

--post-command new "echo {filepath|shell_quote} >> {shell_quote,{export_dir}/exported.txt}"

In the above command, the 'shell_quote' filter is used to ensure '{filepath}' is properly quoted and the
'{shell_quote}' template ensures the constructed path of '{exported_dir}/exported.txt' is properly quoted. If
'{filepath}' is 'IMG 1234.jpeg' and '{export_dir}' is '/Volumes/Photo Export', the command thus renders to:

echo 'IMG 1234.jpeg' >> '/Volumes/Photo Export/exported.txt'

It is highly recommended that you run osxphotos with '--dry-run --verbose' first to ensure your commands are
as expected. This will not actually run the commands but will print out the exact command string which would
be executed.
```

**Reply by nullpointerninja** [reply]
Wow, fantastic! Just tested and it works as you described. From my testing it also keeps the files as clones and not as separate files taking double the space in APFS volumes.
Thanks again!

**Reply by RhetTbull** [reply]
> From my testing it also keeps the files as clones and not as separate files taking double the space in APFS volumes

Yes, osxphotos will use "copy on write" to keep clones on APFS volumes. If you use `--exiftool` though to write metadata you'll end up with double the storage use because the clone will be different than the original.
